saleor/plugins/openid_connect/plugin.py

```diff
@@ -0,0 +1,451 @@
+                }
+            )
+
+    def _get_oauth_session(self):
+        scope = "openid profile email"
+        if self.config.use_scope_permissions:
+            permissions = [f"saleor:{perm}" for perm in get_permissions_codename()]
+            permissions.append("saleor:staff")
+            scope_permissions = " ".join(permissions)
+            scope += f" {scope_permissions}"
+        if self.config.enable_refresh_token:
+            scope += " offline_access"
+        return OAuth2Session(
+            client_id=self.config.client_id,
+            client_secret=self.config.client_secret,
+            scope=scope,
+        )
+
+    def _use_scope_permissions(self, user, scope):
+        user_permissions = []
+        if scope:
+            permissions = get_saleor_permissions_qs_from_scope(scope)
+            user_permissions = get_saleor_permission_names(permissions)
+            user.effective_permissions = permissions
+        return user_permissions
+
+    def external_obtain_access_tokens(
+        self, data: dict, request: WSGIRequest, previous_value
+    ) -> ExternalAccessTokens:
+        if not self.active:
+            return previous_value
+
+        if not self.use_authorization_flow:
+            return previous_value
+
+        code = data.get("code")
+        if not code:
+            msg = "Missing required field - code"
+            raise ValidationError(
+                {"code": ValidationError(msg, code=PluginErrorCode.NOT_FOUND.value)}
+            )
+
+        state = data.get("state")
+        if not state:
+            msg = "Missing required field - state"
+            raise ValidationError(
+                {"state": ValidationError(msg, code=PluginErrorCode.NOT_FOUND.value)}
+            )
+
+        try:
+            state_data = signing.loads(state)
+        except signing.BadSignature:
+            msg = "Bad signature"
+            raise ValidationError(
+                {"state": ValidationError(msg, code=PluginErrorCode.INVALID.value)}
+            )
+
+        redirect_uri = state_data.get("redirectUri")
+        if not redirect_uri:
+            msg = "The state value is incorrect"
+            raise ValidationError(
+                {"code": ValidationError(msg, code=PluginErrorCode.INVALID.value)}
+            )
+
+        token_data = self.oauth.fetch_token(
+            self.config.token_url, code=code, redirect_uri=redirect_uri
+        )
+
+        parsed_id_token = get_parsed_id_token(
+            token_data, self.config.json_web_key_set_url
+        )
+
+        user = get_or_create_user_from_payload(
+            parsed_id_token, self.config.authorization_url
+        )
+
+        user_permissions = []
+        if self.config.use_scope_permissions:
+            scope = token_data.get("scope")
+            user_permissions = self._use_scope_permissions(user, scope)
+            if not user.is_staff and bool("saleor:staff" in scope or user_permissions):
```

Comment:
`saleor:staff` could be moved to some constant. It's being used in a number of places and it's a hardcoded string everywhere.